File: JavaDemoGithubPractice/app/src/main/java/com/example/javademogithubpractice/mvp/presenter/SplashPresenter.java


package com.example.javademogithubpractice.mvp.presenter;

import android.widget.Toast;

import com.example.javademogithubpractice.AppData;
import com.example.javademogithubpractice.mvp.contract.ISplashContract;
import com.example.javademogithubpractice.mvp.model.User;
import com.example.javademogithubpractice.room.DaoSessionImpl;
import com.example.javademogithubpractice.room.model.AuthUser;

import java.util.Date;
import java.util.List;
import javax.inject.Inject;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;


public class SplashPresenter extends BasePresenter<ISplashContract.View> implements ISplashContract.Presenter{

    //private final String TAG = "SplashPresenter";

    private AuthUser authUser;
    private boolean isMainPageShowwed = false;

    private static CompositeDisposable compositeDisposable = new CompositeDisposable();

    private void addDisposable(Disposable disposable) {
        compositeDisposable.add(disposable);
    }

    @Inject
    SplashPresenter(DaoSessionImpl daoSession) {
        super(daoSession);
    }

    @Override
    public void getUser() {
        Observable<List<AuthUser>> observable = daoSession.loadAllAuthUser();
        addDisposable(observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onSuccessAuthUser,this::onErrorUser));
    }


    private void onSuccessAuthUser(List<AuthUser> users) {
        System.out.println("######################GOOOOOOOOOOOOOOOODDD" + users);
        AuthUser selectedUser = users != null && users.size() > 0 ? users.get(0) : null;

        if (selectedUser != null && isExpired(selectedUser)) {
            //authUserDao.delete(selectedUser);
            selectedUser = null;
        }

        if (selectedUser != null) {
            AppData.INSTANCE.setAuthUser(selectedUser);
            getUserInfo(selectedUser.getAccessToken());
        } else {
            mView.showLoginPage();
        }

    }

    @Override
    public void saveAccessToken(String accessToken, String scope, int expireIn) {
        AuthUser authUser = new AuthUser();
        authUser.setSelected(true);
        authUser.setScope(scope);
        authUser.setExpireIn(expireIn);
        authUser.setAuthTime(new Date());
        authUser.setAccessToken(accessToken);
        daoSession.storeAuthUser(authUser);
        this.authUser = authUser;
    }


    private void getUserInfo(final String accessToken) {
        Observable<Response<User>> observable = getUserService(accessToken).getPersonInfo(true);
        addDisposable(observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onSuccessUser,this::onErrorUser));
    }

    private void onSuccessUser(Response<User> response) {
        AppData.INSTANCE.setLoggedUser(response.body());
        if (authUser != null) {
            assert response.body() != null;
            authUser.setLoginId(response.body().getLogin());
            addDisposable(daoSession.storeAuthUser(authUser)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(this::successStore,this::errorStore));
        }
        if(!isMainPageShowwed) {
            isMainPageShowwed = true;
            mView.showMainPage();
        }
    }

    private void errorStore(Throwable throwable) {
        Toast.makeText(getContext(),"Error update" + throwable,Toast.LENGTH_SHORT).show();
    }

    private void successStore() {
        Toast.makeText(getContext(),"Success update",Toast.LENGTH_SHORT).show();
    }

    private void onErrorUser(Throwable throwable) {
        //daoSession.getAuthUserDao().delete(AppData.INSTANCE.getAuthUser());
        AppData.INSTANCE.setAuthUser(null);
        mView.showErrorToast(getErrorTip(throwable));
        mView.showLoginPage();
    }

    private boolean isExpired(AuthUser selectedUser){
        return selectedUser.getAuthTime().getTime() + selectedUser.getExpireIn() * 1000 < System.currentTimeMillis();
    }

    @Override
    public void detachView() {
        compositeDisposable.clear();
        compositeDisposable.dispose();
        mView = null;
    }
}

File: KotlinDemoGithubProject/app/src/main/java/com/example/kotlindemogithubproject/mvp/presenter/SplashPresenter.kt
package com.example.kotlindemogithubproject.mvp.presenter

import androidx.annotation.NonNull
import com.example.kotlindemogithubproject.AppData
import com.example.kotlindemogithubproject.dao.AuthUser
import com.example.kotlindemogithubproject.dao.AuthUserDao
import com.example.kotlindemogithubproject.dao.DaoSession
import com.example.kotlindemogithubproject.http.core.HttpObserver
import com.example.kotlindemogithubproject.http.core.HttpResponse
import com.example.kotlindemogithubproject.mvp.contract.ISplashContract
import com.example.kotlindemogithubproject.mvp.model.User

import javax.inject.Inject
import java.util.Date


class SplashPresenter @Inject constructor(daoSession: DaoSession) : BasePresenter<ISplashContract.View>(daoSession), ISplashContract.Presenter {


    private val TAG = "SplashPresenter"

    private var authUser: AuthUser? = null
    private var isMainPageShowwed = false

    override fun getUser() {
        val authUserDao = daoSession.authUserDao

        val users = authUserDao.queryBuilder()
            .where(AuthUserDao.Properties.Selected.eq(true))
            .limit(1)
            .list()

        var selectedUser: AuthUser? = if (users != null && users!!.size > 0) users!!.get(0) else null

        //if none selected, choose first account
        if (selectedUser == null) {
            val firstAccount = authUserDao.queryBuilder()
                .limit(1)
                .list()
            selectedUser = if (firstAccount != null && firstAccount!!.size > 0) firstAccount!!.get(0) else null
        }

        if (selectedUser != null && isExpired(selectedUser!!)) {
            authUserDao.delete(selectedUser)
            selectedUser = null
        }

        if (selectedUser != null) {
            AppData.INSTANCE.authUser = selectedUser
            getUserInfo(selectedUser.accessToken!!)
        } else {
            mView!!.showLoginPage()
        }

    }

    override fun saveAccessToken(accessToken: String, scope: String, expireIn: Int) {
        val authUser = AuthUser()
        authUser.selected = true
        authUser.scope = scope
        authUser.expireIn = expireIn
        authUser.authTime = Date()
        authUser.accessToken = accessToken
        daoSession.authUserDao.insert(authUser)
        this.authUser = authUser
    }

    private fun getUserInfo(accessToken: String) {
        val httpObserver = object : HttpObserver<User> {
            override fun onError(@NonNull error: Throwable) {
                daoSession.authUserDao.delete(AppData.INSTANCE.authUser)
                AppData.INSTANCE.authUser = null
                getErrorTip(error)?.let { mView!!.showErrorToast(it) }
                mView!!.showLoginPage()
            }

            override fun onSuccess(@NonNull response: HttpResponse<User>) {
                AppData.INSTANCE.loggedUser = response.body()
                if (authUser != null) {
                    authUser!!.loginId =response.body()?.login
                    daoSession.authUserDao.update(authUser)
                }
                if (!isMainPageShowwed) {
                    isMainPageShowwed = true
                    mView!!.showMainPage()
                }
            }
        }

//        generalRxHttpExecute(
//            IObservableCreator<Any> { forceNetWork -> userService.getPersonInfo(forceNetWork) },
//            httpObserver,
//            true
//        )

    }

    private fun isExpired(selectedUser: AuthUser): Boolean {
        return selectedUser.authTime!!.getTime() + selectedUser.expireIn * 1000 < System.currentTimeMillis()
    }

}
